### DISCUSSION TITLE
Next 13/14 - Isomorphic DOMPurify - Build Error

### DISCUSSION THREAD

**Discussion by mikeriley131** [vercel/next.js#58142] [Help]
### Summary

isomorphic-dompurify version - 1.9.0
Node version - 18.18.0
Next.js version - 13.5.4 (also reproduced with Next 14)
React version - 18.2.0

I am importing my company's React component library that makes use of Isomorphic DOMPurify.
No issues when using it in a client-side app.
However, when running yarn build in my Next.js project I am getting ReferenceError: window is not defined and the path in the logs points to the following line of code: var ti=window.DOMPurify||(window.DOMPurify=Vl().default||Vl());

In the issue for the package - https://github.com/kkomelin/isomorphic-dompurify/issues/228 - it was suggested that possibly Next treats the browser field in the package.json as a server entry point. Could this be the culprit?

**Comment by icyJoseph** [comment]
Hi,

Without much knowledge into the matter, it seems to me that the problem is that, this library is doing a strange check... `typeof process === "undefined"` is not a reliable check for client side. They should do `typeof window !== "undefined"`, since after all, they expect to work with the window object.

If you copy their code, and tweak it a bit:

```ts
"use client";

function resolveDOMPurify() {
  const isClientSide = typeof process === "undefined";
  return isClientSide ? console.log("client") : console.log("server");
}

resolveDOMPurify();

export const Button = () => {
  return (
    <button type="button" >
      +1
    </button>
  );
};

``` 

It'll log server on the client.

I am not sure in what context you use it, but that's more or less the problem. 

I am gonna drop something from the Apollo GraphQL Docs:

> It's impossible for Apollo Client to use a guard statement like typeof process !== "undefined" && process.env.NODE_ENV !== 'production' because some bundlers, such as esbuild, cannot replace a statement like typeof process at build time. The development-only code would remain after the build, resulting in no bundle size reduction.
> 
> In contrast, checking for globalThis.__DEV__ does not crash the browser and can be eliminated by every bundler.

As for your issue, perhaps defer usage of this library to the client side, by using next/dynamic, or hiding it behind a boolean that is set on the client only, for example within a use-effect. This of course has implications for the initial HTML string that you'll send to your clients.

**Reply by icyJoseph** [reply]
And if the problem is that the bundler is picking up the `browser` field in the package, then we have to look into the webpack configuration, but that is a bit more tricky I think, we'd probably have to use some resolver field, to tell webpack where to take this library from, when `nextRuntime` is not "undefined"

**Reply by icyJoseph** [reply]
If you could produce a small reproduction demo as well, that'd be nice. Maybe it is something with your specific setup?

**Reply by icyJoseph** [reply]
How are you triggering this issue? I tried here for a bit, but it seems to work? https://stackblitz.com/edit/nextjs-6ystvu?file=app%2Fpage.tsx,app%2FButton.tsx

**Reply by icyJoseph** [reply]
Mmm that Stackblitz template used 13.5.1, and 13.5.4 won't work there 😬 ~ I'll try out some local project I have lying around.

**Reply by mikeriley131** [reply]
Thanks for the responses. The Isomorphic DOMPurify package is being used in my company's component library which is then imported into the Next.js app. The error appears during dev on any page that uses a component from the library or when running `next build`.

The component library is within a private registry, so unfortunately I am not able to create a reproduction that is publicly available.

**Reply by icyJoseph** [reply]
But, shouldn't it fail with DOMPurify somehow? I just tested with Next 14.0.1 importing `isomorphic-dompurify` in server and client components, and the dev and build runs work without issue.

I am most definitely, oversimplifying, or doing it wrong though :/ We need a reproduction demo.

**Reply by mikeriley131** [reply]
I'll see if I can put a repro together this evening. Thanks.

**Reply by kkomelin** [reply]
As for the `typeof process !== "undefined"`, I don't know a better way of checking that. Not sure that [globalThis.__DEV__](https://www.apollographql.com/docs/react/development-testing/reducing-bundle-size/#why-use-globalthis__dev__-instead-of-processenvnode_env) is available/set everywhere. Please correct me if I'm wrong.

**Reply by icyJoseph** [reply]
`typeof window !== 'undefined'` should be enough right? If the aim is to work with the window object.

Without a reproduction repository, this one is gonna be hard to debug. 

@kkomelin do you have a code sample that we could try out?

**Reply by kkomelin** [reply]
@icyJoseph Thanks.

> typeof window !== 'undefined' should be enough right? If the aim is to work with the window object.

It's hard to guarantee that no one actually sets that` window` object to simulate the browser behavior, similar to what bundlers do with `process`.

I don't have reproduction code unfortunately. Let's wait for @mikeriley131 for that.

**Reply by kkomelin** [reply]
@icyJoseph And we need to make sure the `browser` field in `package.json` is not used on the server because it's the root cause of the issue reported here, imo.
The `process `condition has nothing to do with that.

**Reply by icyJoseph** [reply]
The way I see it, we have a bunch of possible routes here, but no reproduction demo. Let's wait until someone can produce one.

**Reply by mikeriley131** [reply]
OK, I was able to re-create it with a faux component library that makes use of isomorphic-dompurify.

Here's a repo for the Next.js app - https://github.com/mikeriley131/next-idp-test
And here's a repo for the component library - https://github.com/mikeriley131/pyrographic-react-component-library

The component library is "npm link"ed to the next app.

When I add the `Button` component from the library to the Next.js app, it throws the following error:
```
Server Error
Error: window is not defined

This error happened while generating the page. Any console logs will be displayed in the terminal window.
Source
../pyrographic-react-component-library/dist/components/Button/index.js (509:9) @ window

  507 | }(Ce)), Ce.exports;
  508 | }
> 509 | var fn = window.DOMPurify || (window.DOMPurify = ht().default || ht());
      |       ^
  510 | function pn(re) {
  511 | const { className: xe, ...z } = re;
  512 | return /* @__PURE__ */ rn("button", { className: `${xe} ${ln.button}`, ...z, children:
  ```

**Reply by kkomelin** [reply]
Thanks for providing the reproduction @mikeriley131 !

I don't know your context fully but from what I see here https://github.com/mikeriley131/pyrographic-react-component-library/blob/main/lib/components/Button/index.tsx#L8, you don't need to sanitize strings enclosed in `{}` because React automatically escapes strings in there. Moreover, you can accidentally remove necessary string parts if you sanitize the string through `isomorphic-dompurify`, for example when the button content has an image or something else.
I would recommend using `isomorphic-dompurify` for the `dangerouslySetInnerHTML` cases primarily instead.

**Reply by kkomelin** [reply]
@mikeriley131 May I ask you to add the reproduction code to the issue description as well?

**Reply by mikeriley131** [reply]
Thanks @kkomelin. This example was just a quick one to verify the issue. My actual use-case is using `sanitize` with `dangerouslySetInnerHTML`.

> I would recommend using dompurify for the dangerouslySetInnerHTML cases primarily instead.

Isn't the point of using `isomorphic-dompurify` is that it will work with server rendered apps, while `dompurify` won't?

**Reply by kkomelin** [reply]
@mikeriley131 Ah, it's just for example, I see.

> Isn't the point of using isomorphic-dompurify is that it will work with server rendered apps, while dompurify won't?

That's true. On "dompurify" here I meant "isomorphic-dompurify". Since "isomorphic-dompurify" is just a wrapper around "dompurify", the sanitization logic works the same way. Will update my comments above not to mislead people.

**Reply by mikeriley131** [reply]
@icyJoseph Any thoughts since seeing the error in the demo repo?

**Reply by icyJoseph** [reply]
Oh, sorry, I thought it was clear. You are bundling your components library incorrectly. 

Forget about Next.js for a moment, and go to the `pyrographic-react-component-library`, build it and inspect the contents of `dist/main.js` I guess? Or maybe it is button or something, but you'll see that Rollup has inlined everything in there, so that when it is imported, say in a Node env, the parser finds a window reference, and it explodes. 

I haven't worked with Rollup for a while now, it ought to be a Vite setting or a Rollup specific, that you have to look at.

### Edit

Yeah you had it:

```console
../pyrographic-react-component-library/dist/components/Button/index.js (509:9) @ window

  507 | }(Ce)), Ce.exports;
  508 | }
> 509 | var fn = window.DOMPurify || (window.DOMPurify = ht().default || ht());
```

It is `dist/components/Button/index.js`, where the bundler has inlined everything. You have to figure out how to create proper bundles for node/browser, etc, through Rollup/Vite, not sure which one is not properly configured.

**Reply by mikeriley131** [reply]
Hmmmm, OK. Thanks for clarifying.
I followed this tutorial - https://dev.to/receter/how-to-create-a-react-component-library-using-vites-library-mode-4lma
I have no idea what the improper configuration might be.

**Reply by icyJoseph** [reply]
Mmm like I said it has been a while since I bundled stuff using rollup, but I think you need to tell it to output two bundles, one for the server and one for clients.

And your package.json should expose a main pointing to the server, and a browser key pointing to the browser bundle.

**Reply by mikeriley131** [reply]
Thank you. I found this post that seems to replicate what you're describing - https://stackoverflow.com/questions/62459725/how-to-import-a-different-bundle-for-server-and-client-using-rollup-in-sapper

I appreciate the help.

**Comment by kkomelin** [comment]
Hey everyone,

The isomorphic-dompurify maintainer is here. 

I do think the [browser.js](https://github.com/kkomelin/isomorphic-dompurify/blob/master/browser.js) is for some reason used here because the code which causes the error is:
```javascript
var ti=window.DOMPurify||(window.DOMPurify=Vl().default||Vl());
```
and the content of the browser.js is:
```javascript
module.exports = window.DOMPurify || (window.DOMPurify = require('dompurify').default || require('dompurify'));
```
which I think are almost identical.

**Reply by BohdanYavorskyi** [reply]
hey @kkomelin so what is the solution for this? Looks like it should have `use client` on top of file so then `window` won't be undefined

**Reply by kkomelin** [reply]
@BohdanYavorskyi Is the `use client `directive supported anywhere but React? The isomorphic-dompurify library is not built for React specifically, so we need to make sure `use client `won't break other environments, where isomorphic-dompurify is used.

**Reply by icyJoseph** [reply]
`use client` doesn't mean browser only. It means that the module will be bundled for a browser/client environment, but it will still be, unless other conditions are met, SSR'd 

- https://github.com/reactwg/server-components/discussions/4

**Reply by BohdanYavorskyi** [reply]
yeah, @icyJoseph is right. So @kkomelin guess we can try it out

**Reply by kkomelin** [reply]
@BohdanYavorskyi you din't answer my question, sorry. it's still valid

**Reply by BohdanYavorskyi** [reply]
@kkomelin @icyJoseph explained you that

**Reply by kkomelin** [reply]
Thanks guys @BohdanYavorskyi @icyJoseph I've released a new version with "use client". Please check it out and let me know if it works for you https://github.com/kkomelin/isomorphic-dompurify/releases/tag/v1.12.0

**Reply by BohdanYavorskyi** [reply]
@kkomelin checked the new version. Still happens :(

I can assume it also fails due to this resolve DOM purify stuff.

In cjs module it just runs when you have any import from the library and it runs on server components where window is `undefined`

**Reply by icyJoseph** [reply]
Mmm I don't think the "use client" directive was necessary. 

Are we talking about the same issue as OP? I remember that OPs case was happening because their library already packed the wrong Dom purify version, so it had nothing to do with Next.js

**Reply by mikeriley131** [reply]
Was it the wrong DOMPurify version? I'm on the latest version. I understood the replies to suggest that I needed to somehow generate a version for SSR apps that does not include references to `window`. But unfortunately have not yet found any guides on how to do that. Did I misunderstand the responses?

**Reply by BohdanYavorskyi** [reply]
@mikeriley131 seems the latest v 3.0.5 is on this isomorphic package. I checked the app 2h ago and it still fails on nextjs app router server pages. 

`I understood the replies to suggest that I needed to somehow generate a version for SSR apps that does not include references to window`

I guess yeah

**Reply by silasabbott** [reply]
This may or may not be related, but I'm experiencing the same error with isomorphic-dompurify when deploying SvelteKit on Vercel Edge Functions. Not sure who's the culprit in these situations. When deploying with Vercel Serverless Functions everything works great.

Here's the [package.json browser field spec](https://github.com/defunctzombie/package-browser-field-spec) for reference. Bundlers/tooling should be respecting this field for browser environments as far as I know.

**Reply by kkomelin** [reply]
I agree with @silasabbott

**Comment by codenoid** [comment]
same with SvelteKit deployed on Cloudflare Pages

**Reply by Le0X8** [reply]
I discovered the same now lol

**Comment by kkomelin** [comment]
Please test the new release of the `isomorphic-dompurify` which addresses the issue and leave your feedback https://github.com/kkomelin/isomorphic-dompurify/releases/tag/2.26.0
